src/mpi/comm/comm_split_type.c

```diff
@@ -173,8 +173,22 @@ static int node_split_processor(MPIR_Comm * comm_ptr, int key, const char *hintv
     if (obj_containing_cpuset->type == query_obj_type) {
         color = obj_containing_cpuset->logical_index;
     } else {
-        hwloc_obj_t hobj = hwloc_get_ancestor_obj_by_type(MPIR_Process.topology, query_obj_type,
-                                                          obj_containing_cpuset);
+        hwloc_obj_t hobj = NULL;
+        if (query_obj_type == HWLOC_OBJ_NUMANODE) {
+            hwloc_obj_t tmp = NULL;
+            while ((tmp =
+                    hwloc_get_next_obj_by_type(MPIR_Process.topology, query_obj_type,
+                                               tmp)) != NULL) {
```

Comment:
I'm not sure I follow the logic here.  When you pass `NULL` to `hwloc_get_next_obj_by_type` in the third parameter, what's the expected return value?

Reply:
hwloc_get_next_obj_by_type is like an iterator that takes NULL initially and returns the first object of the type `query_obj_type`. We need to pass the previous object to move ahead, subsequently.